Fix test images

jsafrane opened the pull request:
These commits fix volume_io tests for iSCSI and Ceph RBD. Both server images were quite old (Fedora 22), so I updated them to ~~something more stable (CentOS 7) and to newer Ceph (Jewel, 10.2.7).~~ something newer (Fedora 26).

The most important fix is that the test volumes have 120 MB so volume_io test can actually run - the tests put 100MB file to the volume to check its persistence.

When mount containers in #53440 are merged I'll try to run the tests regularly with every PR (or merge) so we catch regressions quickly.

```release-note
NONE
```

/sig testing
/sig storage

/assign @jeffvance 

Fixes: #56725

jsafrane commented:
Note that updated images must be merged and pushed to gcr.io before updating versions in [volume_util.go](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/volume_util.go#L60)

I'll do that in subsequent PR when this one is merged.

jsafrane commented:
/assign @rootfs @copejon 
/unassign @jeffvance 
(Jeff is on PTO)

rootfs commented on test/images/volumes-tester/iscsi/Dockerfile:
```diff
@@ -12,8 +12,9 @@
 
-FROM fedora
-RUN yum install -y iscsi-initiator-utils targetcli net-tools strace && yum clean all
+FROM centos
```
set centos release label?

mkumatag commented on test/images/volumes-tester/rbd/Dockerfile:
```diff
@@ -15,10 +15,10 @@
 # CEPH all in one
 # Based on image by Ricardo Rocha, ricardo@catalyst.net.nz
 
-FROM fedora
```
Can you please stick to Fedora? because I have a plan to convert these images to multi-arch like arm, ppc64le etc. So I don't think we really have all these packages available for other archs. As packages are easily available for archs in Fedora, so please shift to latest fedora instead of moving to centos based.

jsafrane replied:
Fedora moves forward too quickly. For example, current test images come from Fedora 22 and nobody has touched them for 2 years. It's not possible to get reproducible build of gcr.io/google_containers/volume-rbd, because "FROM fedora:22" will get you a very different image than the one pushed in gcr.io.

I'd like to stick to something more stable if possible

mkumatag replied:
The only concern I've is that the packages what we are installing part of centos are not available for other archs like arm, ppc64le. So I'm good if we are able to build this image on all the architecture. @detiber, @luxas @ixdy

detiber replied:
@mkumatag @jsafrane: CentOS currently only publishes images for amd64 and armv8. While it would be possible to use a different image for ppc64le and s390x, I think that would have more of a potential to cause confusion and lead to arch-specific bugs/edge cases.

I know @jperrin has been working on adding ppc64le support to the official CentOS image, but that would still leave s390x as the odd arch out.

jsafrane replied:
Ok, I updated the images to Fedora 26 and tested them.

mkumatag commented:
/lgtm for test/images changes

mkumatag reviewed (approved):
/lgtm

mkumatag commented:
Can you please update [ceph  image](https://github.com/kubernetes/kubernetes/blob/master/test/images/volumes-tester/ceph/Dockerfile) also in another PR?

ianchakeres commented on test/images/volumes-tester/iscsi/create_block.sh:
```diff
@@ -33,8 +33,9 @@ cleanup()
+# Create 120MB device with ext2
+# (volume_io tests need at least 100MB)
+dd if=/dev/zero of=block seek=120 count=1 bs=1M
 mkfs.ext2 block
```
While we're touching this script, what do you think about using ext4? Its much more common than ext2.

jsafrane replied:
ext2 fits 1MB image well. Now that we have 120MB, I don't know. Would be the image (in tar.gz format) significantly bigger?

jsafrane commented:
@mkumatag, we track the deps in #53360, these images are really old and need to be updated anyway.

ianchakeres commented on test/e2e/framework/volume_util.go:
```diff
@@ -61,6 +61,8 @@ const (
 	GlusterfsServerImage string = "gcr.io/google_containers/volume-gluster:0.2"
 	CephServerImage      string = "gcr.io/google_containers/volume-ceph:0.1"
 	RbdServerImage       string = "gcr.io/google_containers/volume-rbd:0.1"
```
Should the tags above be updated to the new versions?

jsafrane replied:
No, this PR must be merged first, then someone (from google?) must manually build & push new images and then we can bump versions here. Otherwise we blow all e2e tests.

mkumatag replied:
I agree with @jsafrane that we need to push image first before modifying testcode. @ixdy can you please help us pushing above images so that we can pass these tests with new images?

ixdy replied:
sure, I can help build images, though currently the test-image-build-lock is assigned to #52744.

mkumatag replied:
@ixdy : As #52744 is merged, can you please help us pushing these images?

mkumatag replied:
ping @ixdy

jsafrane commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
/assign @soltysh @eparis
begging for approval

jsafrane commented:
I joined Ceph RBD and CephFS images into one in the last commit. Please re-review (and approve)

jsafrane commented:
/retest

jsafrane commented:
@mkumatag, please review.

rootfs commented on test/e2e/storage/volumes.go:
```diff
@@ -260,7 +260,7 @@ var _ = SIGDescribe("Volumes", func() {
 				Prefix:      "cephfs",
 				ServerImage: framework.CephServerImage,
-				ServerPorts: []int{6789},
+				ServerPorts: []int{6789, 6808},
```
6808 is probably accidental, the osd port ranges 6800:7300
http://docs.ceph.com/docs/hammer/rados/configuration/network-config-ref/

jsafrane replied:
Fixed

rootfs commented:
cross build failed
```
I1114 12:00:06.369] # k8s.io/kubernetes/pkg/kubelet
I1114 12:00:06.370] pkg/kubelet/volume_host.go:144:9: undefined: "k8s.io/kubernetes/pkg/util/mount".NewExecMounter
```

jsafrane commented:
I'm fixing it at https://github.com/kubernetes/kubernetes/pull/55712

rootfs commented:
/lgtm

jsafrane commented:
/retest

jsafrane commented:
/retest

@eparis, can you please approve this PR? Thanks.

jsafrane commented:
/retest

ixdy commented:
Can you rebase this? I can build/push new test images afterwards.

jsafrane commented:
@ixdy, rebased

mkumatag commented:
/lgtm

ixdy commented on test/images/volumes-tester/gluster/Makefile:
```diff
@@ -12,7 +12,7 @@
 # limitations under the License.
 
-TAG = 0.4
+TAG = 0.5
```
there's apparently already a `gcr.io/google-containers/volume-gluster:0.5`, which is exactly the same as 0.4. I guess I'll just push over it.

```
DIGEST        TAGS            TIMESTAMP
5c3e91a365e7  0.4,0.5,latest  2017-01-25T11:02:27
```

ixdy commented on test/images/volumes-tester/rbd/Dockerfile:
```diff
@@ -15,16 +15,17 @@
 
 # Base Packages
-RUN yum install -y wget ceph ceph-fuse strace && yum clean all
+RUN yum install -y wget strace psmisc procps-ng ceph ceph-fuse && yum clean all
```
this is failing, and I'm not sure how to resolve:
```
...
error: policycoreutils-newrole-2.6-6.fc26.x86_64: install failed
...
Error unpacking rpm package policycoreutils-newrole-2.6-6.fc26.x86_64                                      
policycoreutils-newrole-2.6-6.fc26.x86_64 was supposed to be installed but is not! 
```

jsafrane replied:
This is odd, simple `make image` in test/images/volumes-tester/rbd works for me. Can it be a stalled repo? Post whole `make` output.

jsafrane commented:
rebased

@ixdy, can you please retry pushing the images and post whole make output when it fails?

ixdy commented:
Sorry for the delay, this fell off my radar for a bit.

I still am not able to build the rbd image on my corp workstation, but I spun up a GCE VM and it worked there, so :man_shrugging:. I've had similar issues with a few images in the past.

I've now pushed the following:
* `k8s.gcr.io/volume-gluster:0.5`
* `k8s.gcr.io/volume-iscsi:0.2`
* `k8s.gcr.io/volume-rbd:0.2`

(At some point it'd probably be good to move these to gcr.io/kubernetes-e2e-test-images.)

mkumatag commented:
> Sorry for the delay, this fell off my radar for a bit.
> 
> I still am not able to build the rbd image on my corp workstation, but I spun up a GCE VM and it worked there, so 🤷‍♂️. I've had similar issues with a few images in the past.
> 
> I've now pushed the following:
> 
> k8s.gcr.io/volume-gluster:0.5
> k8s.gcr.io/volume-iscsi:0.2
> k8s.gcr.io/volume-rbd:0.2

That's there on my todo list to convert these images to multi-arch and push manifest once they work on x86 platform.
> (At some point it'd probably be good to move these to gcr.io/kubernetes-e2e-test-images.)

mkumatag commented:
thanks @ixdy, @jsafrane can you please update the codebase to pull latest images pushed?

jsafrane committed:
Fix iSCSI image

- create 120MB LUN instead of 1MB for volume_io tests
- increase verbosity
- rebase to Fedora 26
  - targetcli now requires running dbus, so run it

jsafrane committed:
Add grace period to volume_io tests

This speeds up termination of client pods.

jsafrane committed:
Update Gluster image

- rebase to Fedora 26

jsafrane committed:
Add pod cleanup timeout

After client pod is stopped we need to wait for a while for kubelet to
unmount and clean up the client pod before we can destroy the server pod.

Killing e.g. iSCSI server too early may result in stale attached device
that cannot be removed.

jsafrane committed:
Fix Ceph RBD image

- create 120MB volume instead of 1MB for volume_io tests
- rebase to Fedora 26
- added compatibility with ext4 and older ceph clients
- unify CephFS and Ceph RBD images.

eparis commented:
/approve

jsafrane commented:
I pushed updated commits and squashed couple of them. Let's see if it still works, it's 3 months since I tested it last time :-)

jsafrane commented:
e2e tests seem to be happy \o/

jsafrane commented:
@mkumatag, PTAL and lgtm.

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/54300)
<!-- Reviewable:end -->

mkumatag commented:
@ixdy are we covering the `[Feature:Volumes]` tests part of any of the above PR jobs? If not can we run them with this PR?

ixdy commented:
I don't think the `[Feature:Volumes]` tests are running anywhere in CI, either pre- or post-submit.

mkumatag commented:
@jsafrance can you confirm whether these tests passed in your local setup?

jsafrane commented:
Tested on CentOS, all tests passed. My ultimate goal is to run the tests also in some e2e suite with containerized mount utilities or on Ubuntu, but I need the images first...

mkumatag commented:
Review status: 0 of 19 files reviewed at latest revision, 1 unresolved discussion.

---

*[test/utils/image/manifest.go, line 99 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54300#-L7-wi9pWMyvg4hfarOB:-L7-wi9pWMyvg4hfarOC:bx5pta7) ([raw file](https://github.com/kubernetes/kubernetes/blob/b8c5bcf48a8852ba7f859a06ce35eac3d8fffafa/test/utils/image/manifest.go#L99)):*
> ```Go
> 	VolumeISCSIServer   = ImageConfig{gcRegistry, "volume-icsci", "0.1", false}
> 	VolumeGlusterServer = ImageConfig{gcRegistry, "volume-gluster", "0.2", false}
> 	VolumeCephServer    = ImageConfig{gcRegistry, "volume-ceph", "0.1", false}
> ```

Why is this `VolumeCephServer` has been removed?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/54300)*
<!-- Sent from Reviewable.io -->

jsafrane commented:
"Fix Ceph RBD image" commit unifies Ceph RBD and CephFS images, one image now provides both.

mkumatag commented:
/lgtm Thanks @jsafrane

mkumatag commented:
/lgtm

mkumatag commented:
Good work @jsafrane, thanks for the contribution.
